status: update cmd ref (#416)
iterative/dvc.org: static/docs/commands-reference/status.md

## Synopsis

```usage

> usage: dvc status [-h] [-v] [-j JOBS] [--show-checksums] [-q] [-c] [-r REMOTE]

Reviewer: why this change? no need to be precise and copy it exactly from the DVC command line output

Author: I did the same, simply copy-paste from command line output.

Reviewer: _no_ need to copy it from the CLI output.

it makes sense to chage `[-q | -v]` but keep multiple lines to avoid scrolling. Especially on the mobile screen.

                  [-a] [-T] [-d]
                  [targets [targets ...]]

> Show changed states, compare local cache and remote storage.

Reviewer: no need in this change, we don't include it usually. A summary at the very top of the page is enough. Check other command - like checkout fr example.

Author: Okay, I will remove these changes.

documentation: https://man.dvc.org/status

positional arguments:

>   targets               DVC files.

Reviewer: let's keep the previous version, and actually we need to update DVC itself. Can you create a patch for DVC repo with this simple change?

Author: Okay, I will revert this to the original version. Okay sure will patch it in DVC as well.

Author: @shcheklein  I have created a pull request for this in DVC kindly review it. 

https://github.com/iterative/dvc/pull/2111

## Synopsis

```usage
usage: dvc status [-h] [-v] [-j JOBS] [--show-checksums]
                  [-q] [-c] [-r REMOTE] [-a] [-T] [-d]
                  [targets [targets ...]]

positional arguments:

>   targets               Limit the scope to these stage files.

Reviewer: update it to the new version?

Author: @shcheklein  Yes, updated it to newer version was waiting for original DVC repo to be updated. Since it got approved so pushed these changes now kindly review them

Reviewer: @sanidhyamangal I think the DVC patch has been merged. Feel free to update the docs.

## Synopsis

```usage

> usage: dvc status [-h] [-v] [-j JOBS] [--show-checksums]

Reviewer: please, address the @jorgeorpinel 's comment ... no split is required

Reviewer 2: Exact txt below 🙂 

```
usage: dvc status [-h] [-v] [-j JOBS] [--show-checksums] [-q] [-c]
                  [-r REMOTE] [-a] [-T] [-d] [targets [targets ...]]
```

Author: reverted it to the original format, unsplit the list. Please review it.

                  [-q] [-c] [-r REMOTE] [-a] [-T] [-d]

>                   [targets [targets ...]]

Reviewer 2: Hi. Why was this changed? In fact it can all fit in 2 lines:
```console
usage: dvc status [-h] [-v] [-j JOBS] [--show-checksums] [-q] [-c]
                  [-r REMOTE] [-a] [-T] [-d] [targets [targets ...]]
```

Reviewer 2: p.s. I updated this in 5709d16 which is a change in my current PR, #464 so you can ignore this one but will probably cause a conflict before you can merge... Or feel free to also apply the same change in advance to avoid such Git conflict.

In the `local` mode, changes are detected through the checksum of every file
listed in every stage file in the pipeline against the corresponding file in the

> file system, using two modes `changed checksum` when actual checksum of data

Reviewer: these are not modes. These are different outputs/statuses. The paragraph that starts with "If instead differences ..." should be updated. Check the list of different outputs that starts line 57-60.

    Pipeline is up to date. Nothing to reproduce.
```

This indicates that no differences were detected, and therefore no stages would
be run again by `dvc repro`.

If instead, differences are detected, `dvc status` lists those changes. For each
DVC-file (stage) with differences, the changes in _dependencies_ and/or
_outputs_ that differ are listed. For each item listed, either the file name or
the checksum is shown, and additionally a status word is shown describing the
changes. This changes list provides a reference to both the status of a
DVC-file, as well as the changes to individual dependencies and outputs

> described in it:

Reviewer 2: Please end this paragraph in period `.` not column `:`

Reviewer 2: Fixed in d0bbaba

> - For the local workspace:

Reviewer 2: This shouldn't be a bullet with sub-bullets. How about using a `###` header? Or just a **bold** phrase.

Reviewer 2: ...and un-indent the bullets below, of course 🙂

    (orphans), which is considered always changed;
  - _changed deps_ means that some changes in dependencies are incorporated in
    the DVC-file, all these states are:
    - _new_ : when reference of new dependency files are added into DVC-file;
    - _modified_ : when checksum of an exsisting dependencies are changed in
      workspace but not reflected in the DVC-file;
    - _deleted_ : when an exsisting dependencies are removed from the workspace
      but still referred in the DVC-file;
    - _not in cache_ : when dependencies mentioned in DVC-file no longer exsist
      in local cache;
  - _changed outs_ means that some changes in output states are incorporated in
    the DVC-file, all these states are:
    - _new_ : when reference of new output files are are added into DVC-file;
    - _modified_ : when checksum of an exsisting outputs are changed in
      workspace but not reflected in the DVC-file;
    - _deleted_ : when an exsisting outputs are removed from the workspace but
      still referred in the DVC-file;
    - _not in cache_ : when outputs mentioned in DVC-file no longer exsist in
      local cache;

> - For comparison against a remote cache:

Reviewer 2: Same as https://github.com/iterative/dvc.org/pull/416#discussion_r307958847

  - _new_ means the file exists in the local cache but not the remote cache;
  - _deleted_ means the file doesn't exist in the local cache, but exists in the

>     remote cache;

Reviewer 2: In all of the bullets above (starting in line 64), use consistent formatting:

- [x] All bullets already start with _italic_ term
- [ ] All bullets should either have column `: ` after the term OR the word ` means `, and then the description. Actually I vote for ` - ` instead, like in all the list of the existing command references.
- [ ] All bullets should be full sentences and end in period `.`  We no longer use semicolons `;` for this.

Thanks

Reviewer 2: Also fixed in d0bbaba

```

This says that no differences were detected, and therefore that no stages would

> be run again by `dvc repro`.

Reviewer 2: - [ ] `This says` -> `This indicates`.
- [ ] `therefore that no` -> `therefore no`

Author: amended, kindly review these changes.

If instead, differences are detected, `dvc status` lists those changes. For each
DVC-file (stage) with differences, the changes in _dependencies_ and/or
_outputs_ that differ are listed. For each item listed, either the file name or
the checksum is shown, and additionally a status word is shown describing the
changes in checksum or status. Given list provides a reference to both, the

> status of the DVC-file and differences in the DVC-files:

Reviewer: `differences in the DVC-files` -> `and changes to individual dependencies and outputs described by the file`

`status of the DVC-file` -> `status of a DVC-file`

Reviewer 2: Here, full suggestion for the last 2 sentences:

"For each item listed, either the file name or the checksum is shown, and additionally a status word is shown describing the changes. This changes list provides a reference to both the status of a DVC-file, as well as the changes to individual dependencies and outputs described in it:"

Author: revised the para based on these suggestions, kindly check them.

Reviewer 2: Yes, thanks. But actually we should also end the paragraph in period `.` please (since there are columns `:` in the main bullets below. (Like I mentioned in https://github.com/iterative/dvc.org/pull/416#discussion_r307958841)

- For the local workspace:

  - _changed checksum_ means actual checksum of the files present in workspace

>     doesn't match the one saved in DVC-files;

Reviewer: it deserves a better explanation tbh. Also, `files` -> `the file`, `in DVC-files` -> `in the DVC-file` ... no need to mention workspace here. The rule of thumb - would you understand anything from this description if you were reading this first time?

Reviewer 2: How about "means that the update checksum of the data doesn't match the one saved in the DVC-files" as I suggested in https://github.com/iterative/dvc.org/pull/416#pullrequestreview-257759178?

Author: updated it in my latest commit, kindly review them.

  - _always changed_ means that this is a special DVC-file with no dependencies

>     (orphans), which is considered always changed;

Reviewer 2: Note that "special DVC-file with no dependencies" is also known as an "orphan" stage file, as mentioned in a few docs:

![image](https://user-images.githubusercontent.com/1477535/60850320-09a89200-a1ac-11e9-837f-4d9ceb17136d.png)

Should we mention that term here? (@shcheklein)

  - _changed deps_ means that there are some changes in dependencies that are

>     incorporated in the DVC-file, some of the states are:

Reviewer: `some` - it's not `some`, it should be a complete list

Author: Paraphrased the given line, please review them.

>     - _new_ : when reference of new dependency files are added into DVC-file;

Reviewer: @efiop just to double check. is it a correct description of this status? can you take a brief look at other statuses as well please.

    - _modified_ : when md5 of an exsisting dependencies are changed in
      workspace but not reflected in the DVC-file;
    - _deleted_ : when an exsisting dependencies are removed from the workspace
      but still referred in the DVC-file;
    - _not in cache_ : when dependencies mentioned in DVC-file no longer exsists
      in local cache;
  - _changed outs_ means that there are some changes in output states that are

>     incorporated in the DVC-file, some of the states are:

Reviewer: `some` - not some, it should be a complete list

Author: Paraphrased for this line as well kindly review the changes.

    - _new_ : when reference of new output files are are added into DVC-file;

>     - _modified_ : when md5 of an exsisting outputs are changed in workspace but

Reviewer: if you mention checksum everywhere above, let's use checksum here as well, instead of md5

Author: Done!

      not reflected in the DVC-file;
    - _deleted_ : when an exsisting outputs are removed from the workspace but
      still referred in the DVC-file;

>     - _not in cache_ : when outputs mentioned in DVC-file no longer exsists in

Reviewer: exists -> exist

> be run again if `dvc repro` were executed.

Reviewer 2: Please improve grammar to: "...no stages would be run again by `dvc repro`."

If instead, differences are detected, `dvc status` lists those changes. For each

> DVC-file (stage) with differences, the changes in both _dependencies_ and/or

Reviewer 2: Can't use "both" with "and/or". I'd remove "both" since "and/or" is more correct.

_outputs_ that differ are listed. For each item listed, either the file name or

> the checksum is shown, and additionally, a status word is shown describing the

Reviewer 2: I don't see the need for that extra comma after `and additionally`, consider reverting.

changes in checksum or status for both local and cloud workspace. Given list

> provides a reference to both, the status of the DVC-file and differences in the

Reviewer: `the status of the DVC-file as well as dependencies and outputs:`

Reviewer 2: What is a "cloud workspace"?
These last 2 sentences in the paragraph are confusing. They seem unnecessarily long and maybe repetitive. Feel free to break up the paragraph in 2 if needed or even re-write the whole thing from scratch, as long as the explanation is clear.

Reviewer 2: Moved to https://github.com/iterative/dvc.org/pull/416/files#r303261384

DVC-files:

- For the local workspace:

  - _changed checksum_ means actual checksum of the DVC-file doesn't match the

>     one specified in it in DVC-files

Reviewer: fix "in it in"

Reviewer 2: "means that the update checksum of the data doesn't match the one saved in the DVC-files."

>   - _always changed_ means that this is a special DVC-file with no dependencies,

Reviewer: for all the items, there should be some rule about punctuation at the end. Most likely ; will do. @jorgeorpinel do you know about this?

Reviewer 2: I haven't standardized this yet but actually I think we mostly use `.` throughout the docs since our bullet lists tend to contain long paragraphs.

Reviewer 2: Also on this line, note that "special DVC-file with no dependencies" is also known as an "orphan" stage file, as mentioned in a few docs:

![image](https://user-images.githubusercontent.com/1477535/60850320-09a89200-a1ac-11e9-837f-4d9ceb17136d.png)

Reviewer: good points. I think the rule should be - if items are short one sentence long, and start from lowercase then we should be using `;` (or nothing?), otherwise `.` make sense.

Reviewer 2: OK, I just standardized it in 3134296. Long list items (paragraphs) and all cmd ref/ config option list items end in period `.` and the list items are separated by a new line. Shorter lists (a sentence or less) end in semicolon `;` and don't have extra new lines in between. I also fixed the indentation of a bunch of lists.

The part about starting in lower case varies though. All the cmd ref option list items start in lower case even when they tend to be full paragraphs for ex., since they are preceded by a `-` dash. E.g.

```
- `-q`, `--quiet` - do not write anything to standard output. Exit with 0 if no
  problems arise, otherwise 1.
```

    which is considered always changed
  - _changed deps_ means that there are some changes in dependencies that are

>     incorporated in the DVC-file, some of the prominent states are:

Reviewer: please, remove prominent. It's a complete list, no need to put extra words on top

>     - _new_ : when new dependencies are added into DVC-file

Reviewer: in this and the following items: it's more about the state of a file or a directory that is dependency, not about changes in the DVC-file. Let's discuss this on Discrord.

    - _modified_ : when an exsisting dependencies are modified in DVC-file
    - _deleted_ : when an exsisting dependencies are removed from DVC-file

>     - _not in cache_ : when dependencies mentioned in DVC-file no longer exsists

Reviewer: no longer - it can be not yet

let's just keep "... don't exist in DVC cache"

Reviewer: mind the comment above ^^

If instead, differences are detected, `dvc status` lists those changes. For each
DVC-file (stage) with differences, the _dependencies_ and/or _outputs_ that
differ are listed. For each item listed, either the file name or the checksum is
shown, and additionally a status word is shown describing the changes in
checksum or status for both local and cloud workspace:

- For the local workspace:

>   - _changed checksum_ means actual checksum of the DVC-file doesn't match the

Reviewer: can you show an example how does it look like? The same for always changed and changed checksum. Let's artificially create a few example and see how does it look like, please.

there is still some disconnect in this list, in what is described in the paragraph above.

Specifically,  _changed_, _changed checksum_, and _ always changed_ are statuses of the DVC-file itself.

Author: @shcheklein shouldn't we add some examples in simple usage section?

Reviewer: that would be ideal, sure!

Author: @shcheklein I have added the examples in the simple usage section kindly check them out.

Reviewer: this is cool. So, you see:

```
test.dvc:
	changed outs:
		modified:           test
	changed checksum
	always changed
```

Items like `modified` are actually not on the same level of hierarchy as `always changed`.

Why don't we come with a better description that explains these two lists separately. And explains the connection between them.

    one specified in it in DVC-files
  - _always changed_ means that this is a special DVC-file with no dependencies,
    which is considered always changed

>   - _new_ output is displayed when a new dependeny or output state is added to

Reviewer: remove first output - it's confusing since we use output in two different meanings in the same sentence

Reviewer: same below

If instead, differences are detected, `dvc status` lists those changes. For each
DVC-file (stage) with differences, the _dependencies_ and/or _outputs_ that
differ are listed. For each item listed, either the file name or the checksum is
shown, and additionally, a status word is shown describing the changes in

> checksum or status for both local and cloud workspace. Given list describes

Reviewer: Given list sounds like a reference to the listed items above, not to the list below

both, the status of the DVC-file and differences in the DVC-files:

- For the local workspace:

  - _changed checksum_ means actual checksum of the DVC-file doesn't match the
    one specified in it in DVC-files
  - _always changed_ means that this is a special DVC-file with no dependencies,
    which is considered always changed
  - _new_ is displayed when a new dependeny or output state is added to DVC-file

>   - _modified_ is displayed when any dependencies or output states is modified

Reviewer: dependencies ... are
dependency ... is

    in DVC-file

>   - _deleted_ is displayed when any exsisting dependencies or output states are

Reviewer: same here

differ are listed. For each item listed, either the file name or the checksum is
shown, and additionally a status word is shown describing the change:

> - For the changes detected in local workspace:

Reviewer: it's not clear how this status is connected with the deps/outs statues below. It looks they are two separate lists which is not the case. One of them encloses the other.

Author: @shcheklein, I have merged these two lists into the previous heading this should work now. Kindly review the changes.

  - _changed deps_ means the DVC-file has changed dependencies
  - _changed outs_ means the DVC-file has changed outputs
  - _changed checksum_ means actual checksum of the DVC-file doesn't match the
    one specified in it in DVC-files
  - _always changed_ means that this is a special DVC-file with no dependencies,
    which is considered always changed

> - For the output displayed in local workspace:

Reviewer: is it the same for dependencies?

Author: Yes, it is the same for dependencies too.

shown, and additionally a status word is shown describing the change:

- For the local workspace:

>   - _changed deps_ means the named file has changed dependencies

Reviewer 3: ```suggestion
  - _changed deps_ means the DVC-file has changed dependencies
```

>   - _changed outs_ means the named file has changed outputs

Reviewer 3: ```suggestion
  - _changed outs_ means the DVC-file has changed outputs
```

>   - _changed checksum_ means actual check sum of data dosen't match specified in

Reviewer: DVC-files

Reviewer: checksum

Reviewer: `of data` -> of the file

dosen't -> doesn't

>   - _changed checksum_ means actual checksum of the file doesn't match specified

Reviewer 3: ```suggestion
  - _changed checksum_ means actual checksum of the DVC-file doesn't match the one specified in it
```

>     dvc files

Reviewer: in the DVC-file

>   - _always checksum_ means special dvc file with no dependencies, considered

Reviewer: DVC-file

Reviewer: is it `always checksum` or something else? cc @efiop

Reviewer 3: It is something else, for sure. Probably `always changed`.

>   - _always checksum_ means special dvc file with no dependencies, considered

Reviewer 3: ```suggestion
  - _always changed_ means that this is a special DVC-file with no dependencies, considered
```

>     always changed

Reviewer: @efiop could you take a look please? do we mention all possible variants now?

Reviewer 3: We are missing statuses for the out/dep files themselves. E.g. 
```
not in cache
deleted
modified
new
```

Author: @efiop shall I add these statues too in the description?

Reviewer 3: @sanidhyamangal Yes, please :slightly_smiling_face:

>     always changed

Reviewer 3: ```suggestion
```

shown, and additionally a status word is shown describing the change:

- For the local workspace:
  - _changed deps_ means the DVC-file has changed dependencies
  - _changed outs_ means the DVC-file has changed outputs
  - _changed checksum_ means actual checksum of the DVC-file doesn't match the
    one specified in it in DVC-files
  - _always changed_ means that this is a special DVC-file with no dependencies,
    which is considered always changed
  - _new_ means the DVC-file has been created but not added into local cache
  - _modified_ means the dependecies and output is changed in the DVC-file
  - _deleted_ means the DVC-file has been deleted from local cache

>   - _not in cache_ means the DVC-file no longer exist in local cache

Reviewer: so, this is not exactly correct presentation of those states.

first 4 mean that state of the file itself

last 4 mean the state of the output or dependency

We need to change the way we present them.

Please, take the get-started repo, do some simple experiments to get those states, see the pattern.

```dvc
$ dvc status

bar.dvc:

>         changed deps:

Author: @shcheklein, Check this I have already split the _changed_ into _deps_ and _outs_

```dvc
$ dvc status

bar.dvc:
        changed deps:

>                 changed:      bar

Reviewer 3: This is an old status. Now we provide more info. See https://github.com/iterative/dvc.org/pull/416/files#r293507445

Author: @efiop  so shall I update this changed with any of these random statuses or do you want any particular one?

Reviewer 3: @sanidhyamangal having multiple would be great. Thanks! :slightly_smiling_face:

Author: @efiop  okay cool, I will update them.
